### PR TITLE
ARROW-11188: [Rust] Support crypto functions from PostgreSQL dialect

### PR DESCRIPTION
**Pull request by ovr**
Implemented functions:

- [x] MD5 (return type string)
- [x] SHA224 (return type binary)
- [x] SHA256 (return type binary)
- [x] SHA384 (return type binary)
- [x] SHA512 (return type binary)

### CODE REVIEW

**Comment by alamb** [comment]
This PR probably needs to be rebased to pick up the fix for #9138 FYI

**Comment by ovr** [comment]
Thank you @alamb for notice 👍 

 I've done with PR, marked it as ready for review and awaiting review from DF's team.

**Review by jorgecarleitao — COMMENTED** [review]
@ovr , thanks a lot, looks really good. 👍  I left some minor comments. :)

File: rust/datafusion/src/physical_plan/crypto_expressions.rs
Comment (ovr) [reply rust/datafusion/src/physical_plan/crypto_expressions.rs]: Yes, it will crash.

Replaced this with `Ok(array.iter().map(|x| x.map(|x| $FUNC(x))).collect())` without as_slice
And it works. Weird, Where `SHA2DigestOutput<D>` converted to slice.

File: rust/datafusion/src/physical_plan/crypto_expressions.rs
Comment (ovr) [reply rust/datafusion/src/physical_plan/crypto_expressions.rs:106]: Good catch, I added tests for nulls and empty strings by SQL execution.

File: rust/datafusion/tests/sql.rs
Comment (ovr) [inline rust/datafusion/tests/sql.rs:1878]: Tested with real PostgreSQL.

```sql
select
        md5('tom') AS md5_tom,
        md5('') AS md5_empty_str,
        md5(null) AS md5_null,
        encode(sha224('tom'), 'hex') AS sha224_tom,
        encode(sha224(''), 'hex') AS sha224_empty_str,
        sha224(null) AS sha224_null;
```

```json
[
  {
    "md5_tom": "34b7da764b21d298ef307d04d8152dc5",
    "md5_empty_str": "d41d8cd98f00b204e9800998ecf8427e",
    "md5_null": null,
    "sha224_tom": "0bf6cb62649c42a9ae3876ab6f6d92ad36cb5414e495f8873292be4d",
    "sha224_empty_str": "d14a028c2a3a2bc9476102bb288234c415a2b01f828ea62ac5b3e42f",
    "sha224_null": null
  }
]
```

**Comment by ovr** [comment]
Thank you, @jorgecarleitao, for your review. I've added tests + fix handling null values.

I've compared it with PostgreSQL, and It works as expected.

```sql
select
        md5('tom') AS md5_tom,
        md5('') AS md5_empty_str,
        md5(null) AS md5_null,
        encode(sha224('tom'), 'hex') AS sha224_tom,
        encode(sha224(''), 'hex') AS sha224_empty_str,
        sha224(null) AS sha224_null;
```

```json
[
  {
    "md5_tom": "34b7da764b21d298ef307d04d8152dc5",
    "md5_empty_str": "d41d8cd98f00b204e9800998ecf8427e",
    "md5_null": null,
    "sha224_tom": "0bf6cb62649c42a9ae3876ab6f6d92ad36cb5414e495f8873292be4d",
    "sha224_empty_str": "d14a028c2a3a2bc9476102bb288234c415a2b01f828ea62ac5b3e42f",
    "sha224_null": null
  }
]
```

Thanks

File: rust/datafusion/src/physical_plan/functions.rs
Comment (alamb) [inline rust/datafusion/src/physical_plan/functions.rs:502]: 👍  nice cleanup

**Review by alamb — APPROVED** [review]

File: rust/datafusion/Cargo.toml
Comment (jorgecarleitao) [reply rust/datafusion/Cargo.toml:64]: I generally agree with you, @alamb. In this case, we want to support posgres dialect, so it makes sense to support these functions (and not implement these ourselves, as they are even security related).

In general, as long as the crates are small, I do not see a major issue. Our expensive dependencies are Tokio, crossbeam, etc, specially because they really increase the compile time (e.g. compared to the arrow crate).

We already offer a scalar UDF that has the same performance as our own expressions. So, I think that this is the most we can do here.

**Comment by mqy** [comment]
Perhaps the PR description should be rewrite before merge.

**Comment by alamb** [comment]
I filed https://issues.apache.org/jira/browse/ARROW-11214 to track the feature flag idea

**Comment by alamb** [comment]
Thanks again for the contribution @ovr !

**Pull request closed without merging**
